From fd1327e45ee04b2e6e267d4e844a588966bfea54 Mon Sep 17 00:00:00 2001
From: Lucas Morishita <lucas.morishita@toradex.com>
Date: Mon, 31 Jul 2023 09:30:15 -0300
Subject: [PATCH] ostree-fetcher-curl: set max parallel connections

Allow the maximum number of parallel downloads to
be set by an environment variable. The code was
heavily based on this pull request:
https://github.com/uptane/meta-updater/pull/29

Upstream-Status: Inappropriate [needs upstream discussion]

Signed-off-by: Lucas Morishita <lucas.morishita@toradex.com>
Signed-off-by: Jeremias Cordoba <jeremias.cordoba@toradex.com>
---
 src/libostree/ostree-fetcher-curl.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/libostree/ostree-fetcher-curl.c b/src/libostree/ostree-fetcher-curl.c
index fdf8a2ef..7144f3ef 100644
--- a/src/libostree/ostree-fetcher-curl.c
+++ b/src/libostree/ostree-fetcher-curl.c
@@ -230,8 +230,14 @@ _ostree_fetcher_init (OstreeFetcher *self)
   curl_multi_setopt (self->multi, CURLMOPT_TIMERFUNCTION, update_timeout_cb);
   curl_multi_setopt (self->multi, CURLMOPT_TIMERDATA, self);
 #if CURL_AT_LEAST_VERSION(7, 30, 0)
+  long curlm_max_total_conn = 0L;
+  const char* curlm_max_total_conn_str = g_getenv ("OSTREE_CURLM_MAX_TOTAL_CONN");
+  if (curlm_max_total_conn_str != NULL)
+    curlm_max_total_conn = atoi(curlm_max_total_conn_str);
   /* Let's do something reasonable here. */
-  curl_multi_setopt (self->multi, CURLMOPT_MAX_TOTAL_CONNECTIONS, 8);
+  if (curlm_max_total_conn == 0)
+      curlm_max_total_conn = 8;
+  curl_multi_setopt (self->multi, CURLMOPT_MAX_TOTAL_CONNECTIONS, curlm_max_total_conn);
 #endif
   /* This version mirrors the version at which we're enabling HTTP2 support.
    * See also https://github.com/curl/curl/blob/curl-7_53_0/docs/examples/http2-download.c
-- 
2.41.0
